sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/DefaultSource.scala

```diff
@@ -91,6 +96,70 @@ class DefaultSource extends FileFormat with DataSourceRegister {
     new CSVOutputWriterFactory(csvOptions)
   }
 
+  override def buildReader(
+      sqlContext: SQLContext,
+      physicalSchema: StructType,
+      partitionSchema: StructType,
+      dataSchema: StructType,
+      filters: Seq[Filter],
+      options: Map[String, String]): (PartitionedFile) => Iterator[InternalRow] = {
+    val csvOptions = new CSVOptions(options)
+    val headers = dataSchema.fields.map(_.name)
+
+    val conf = new Configuration(sqlContext.sparkContext.hadoopConfiguration)
+    val broadcastedConf = sqlContext.sparkContext.broadcast(new SerializableConfiguration(conf))
+
+    (file: PartitionedFile) => {
+      val fileSplit = {
+        val filePath = new Path(new URI(file.filePath))
+        new FileSplit(filePath, file.start, file.length, Array.empty)
+      }
+
+      val hadoopAttemptContext = {
+        val conf = broadcastedConf.value.value
+        val attemptID = new TaskAttemptID(new TaskID(new JobID(), TaskType.MAP, 0), 0)
+        new TaskAttemptContextImpl(conf, attemptID)
+      }
+
+      val reader = new LineRecordReader()
+      reader.initialize(fileSplit, hadoopAttemptContext)
+
+      val lineIterator = new RecordReaderIterator(reader).map { line =>
+        new String(line.getBytes, 0, line.getLength, csvOptions.charset)
+      }
```

Comment:
`RecordReaderIterator` instantiation can be simplified using [`HadoopFileLinesReader`][1] introduced in PR #11960.

[1]: https://github.com/apache/spark/pull/11960/files#diff-c0332718b0741d1b84c7157120b87888R33